### ISSUE TITLE
Layers being skipped during static quantization

### ISSUE THREAD

**Issue by Rabie-H** [pytorch/pytorch#109612]
### 🐛 Describe the bug

Some layers are being deleted after static quantization: `Flatten` and `BatchNorm` even for blocks that are excluded from quantization.

Torch Version: 2.0.1
Model obtained from: https://github.com/ZhaoJ9014/face.evoLVe

* Code used for quantization

Note that the `output_layer` is excluded from quantization

``` python
import torch
from torch.ao.quantization import get_default_qconfig
from torch.ao.quantization.quantize_fx import prepare_fx, convert_fx
from torch.ao.quantization import QConfigMapping
from tqdm import tqdm

model_.eval()
qconfig = get_default_qconfig("qnnpack")
qconfig_mapping = QConfigMapping().set_global(qconfig)

qconfig_dict = {
    "": qconfig,  
    "module_name": [("output_layer", None)],
}

example_inputs = torch.rand(1, 3, 112, 112).cpu()  
prepared_model = prepare_fx(model_, qconfig_dict, example_inputs)
calibrate(prepared_model, data, steps=200) 
quantized_model = convert_fx(prepared_model)
```

* Block before quantization:
```
(23): bottleneck_IR(
      (shortcut_layer): MaxPool2d(kernel_size=1, stride=1, padding=0, dilation=1, ceil_mode=False)
      (res_layer): Sequential(
        (0): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (2): PReLU(num_parameters=512)
        (3): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (4): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True) =========================> this layer is deleted
      )
```

* Block after quantization:

```
(23): Module(
      (shortcut_layer): MaxPool2d(kernel_size=1, stride=1, padding=0, dilation=1, ceil_mode=False)
      (res_layer): Module(
        (0): QuantizedBatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): QuantizedConv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), scale=0.07121843099594116, zero_point=165, padding=(1, 1), bias=False)
        (2): QuantizedPReLU()
        (3): QuantizedConv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), scale=0.13347403705120087, zero_point=127, padding=(1, 1))
      )
```
The last `BatchNorm2d` layer is gone.

* `output_layer` before quantization

```
(output_layer): Sequential(
    (0): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (1): Dropout(p=0.5, inplace=False)
    (2): Flatten()
    (3): Linear(in_features=25088, out_features=512, bias=True)
    (4): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True) =========================> this layer is deleted
  )
```

* `output_layer` after quantization

```

(output_layer): Module(
    (0): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (1): Dropout(p=0.5, inplace=False)
    (3): Linear(in_features=25088, out_features=512, bias=True)
  )

```

### Versions

```
Collecting environment information...
PyTorch version: 2.0.1+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-6ubuntu2) 7.5.0
Clang version: Could not collect
CMake version: version 3.27.2
Libc version: glibc-2.31

Python version: 3.10.8 (main, Nov 24 2022, 14:13:03) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.8.0-55-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.3.109
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 2080 Ti
Nvidia driver version: 465.27
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/local/cuda-11.3.1/targets/x86_64-linux/lib/libcudnn.so.8.5.0
/usr/local/cuda-11.3.1/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.5.0
/usr/local/cuda-11.3.1/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.5.0
/usr/local/cuda-11.3.1/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.5.0
/usr/local/cuda-11.3.1/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.5.0
/usr/local/cuda-11.3.1/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.5.0
/usr/local/cuda-11.3.1/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.5.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   43 bits physical, 48 bits virtual
CPU(s):                          12
On-line CPU(s) list:             0-11
Thread(s) per core:              2
Core(s) per socket:              6
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       AuthenticAMD
CPU family:                      23
Model:                           8
Model name:                      AMD Ryzen 5 2600 Six-Core Processor
Stepping:                        2
Frequency boost:                 enabled
CPU MHz:                         1767.684
CPU max MHz:                     3400,0000
CPU min MHz:                     1550,0000
BogoMIPS:                        6799.20
Virtualization:                  AMD-V
L1d cache:                       192 KiB
L1i cache:                       384 KiB
L2 cache:                        3 MiB
L3 cache:                        16 MiB
NUMA node0 CPU(s):               0-11
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Full AMD retpoline, IBPB conditional, STIBP disabled, RSB filling
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb hw_pstate sme ssbd sev ibpb vmmcall fsgsbase bmi1 avx2 smep bmi2 rdseed adx smap clflushopt sha_ni xsaveopt xsavec xgetbv1 xsaves clzero irperf xsaveerptr arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif overflow_recov succor smca

Versions of relevant libraries:
[pip3] facenet-pytorch==2.2.9
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.23.5
[pip3] torch==2.0.1
[pip3] torchvision==0.15.2
[pip3] triton==2.0.0
[conda] facenet-pytorch           2.2.9                    pypi_0    pypi
[conda] numpy                     1.23.5                   pypi_0    pypi
[conda] torch                     2.0.1                    pypi_0    pypi
[conda] torchvision               0.15.2                   pypi_0    pypi
[conda] triton                    2.0.0                    pypi_0    pypi
```

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @leslie-fang-intel

**Comment by HDCharles** [comment]
those ops are getting fused since they're both affine

something like:
i.e. ((x*w+b)-mean) / dev = x*[w/dev] + [(b-mean)/dev]

you can check yourself:

```
import torch
from torch.ao.quantization import get_default_qconfig, QConfigMapping
from torch.ao.quantization.quantize_fx import convert_fx, prepare_fx
from tqdm import tqdm

class Net(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.res_layer = torch.nn.Sequential(
            torch.nn.Conv2d(3, 3, 1),
            torch.nn.BatchNorm2d(3),
        )
        self.output_layer = torch.nn.Sequential(
            torch.nn.BatchNorm2d(3),
            torch.nn.Dropout(0),
            torch.nn.Flatten(),
            torch.nn.Linear(300, 3),
            torch.nn.BatchNorm1d(3),
        )

    def forward(self, x):
        x = self.res_layer(x)
        return x

x = torch.randn(10, 3, 10, 10)
model_ = Net()
model_.res_layer[0].weight.data = torch.ones(3,3,1,1) 
model_(x)
model_(x*.5)

model_.eval()

y_normal = model_(x)
y_without_bn = model_.res_layer[0](x)

print("diff with/without bn", (y_normal-y_without_bn).abs().sum())

qconfig = get_default_qconfig("qnnpack")
qconfig_mapping = QConfigMapping().set_global(qconfig)

qconfig_dict = {
    "": qconfig,
    "module_name": [("output_layer", None)],
}
# [(x*a+b-m)/s]
# x*a/s + (b-m)/s

print("expected folded weight", (model_.res_layer[0].weight/model_.res_layer[1].running_var**(1/2))[0].flatten())
prepared_model = prepare_fx(model_, qconfig_dict, x)
print("weight after prepare", getattr(prepared_model.res_layer, "0").weight.flatten())
y_after = prepared_model(x)

print("diff before/after prepare", (y_normal - y_after).abs().sum())

quantized_model = convert_fx(prepared_model)
```

diff with/without bn tensor(348.4685, grad_fn=<SumBackward0>)
expected folded weight tensor([0.9362, 0.9362, 0.9362, 0.9362, 0.9362, 0.9362, 0.9362, 0.9362, 0.9362],
       grad_fn=<ViewBackward0>)
weight after prepare tensor([0.9362, 0.9362, 0.9362, 0.9362, 0.9362, 0.9362, 0.9362, 0.9362, 0.9362],
       grad_fn=<ViewBackward0>)
diff before/after prepare tensor(0.0002, grad_fn=<SumBackward0>)
